### DISCUSSION TITLE
Lattice2 and KicadStepUp Model Positions

### DISCUSSION THREAD

**Discussion by jerseyp2** [DeepSOIC/Lattice2#69] [General]
Hello

I am trying to use Lattice2 and KicadStepUp to position 300 LEDs along a circle/ellipse in FreeCad.

I've used Lattice2 to create a circle of 300 placements, but I cannot figure out how to attach the KicadStepUp models to them in an ordered sequence.

In Kicad 6.99 nightly (most recent) I created a schematic with 300 LEDs, and in PCBnew, I used the default import positions after updating from the schematic. I tried to use Kicad's polar array tool, but the process never finishes after being initiated.

After importing the Kicad board into FreeCad, what I would like to do is arrange the 300 LEDs around a circle, and then push the positions and the board edge back to Kicad. I've watched quite a few videos on Lattice2, and I've tried a variety of things, but I admit I'm stumped.

The order of the LEDs is important to me, I'm trying to position D1-D300 along the circle from a known starting position. They are ordered randomly in the FreeCad tree, and I was trying to avoid going in and giving each of the 300 LEDs its own placement. It may not be possible to use Lattice2 and KicadStepUp together to achieve this, so I might have to place everything manually. Lattice2 seems so powerful that I suspect it's my understanding of how these might work together.

Any information greatly appreciated.

**Comment by DeepSOIC** [comment]
Hi! Unfortunately i don't know how KicadStepUp operates, so i can't really tell you if it is even possible. Lattice2 must be able to create copies of 3d shapes placed as you want, but it cannot (over)write placements of existing objects at the moment (freecad architecture does not allow to do it in a parametric way).

I'm afraid, may have to write a script to do what you need. If you are ready, this is the line of code to obtain the list of placements from Lattice2 array:
```
plms = [sh.Placement for sh in App.ActiveDocument.PolarArray.Shape.childShapes()] # replace "PolarArray" with the name of the array object in your project
```
And then you write them with something like
```
for i in range(len(plms)):
  App.ActiveDocument.getObject("LED"+str(i)).Placement = plms[i]
```

**Reply by jerseyp2** [reply]
Hello Thank you very much for the fast reply. I had wondered if it might require scripting. I'm not familiar with Python, but it's worth a try, so I will see what I can do with your code examples, and perhaps I might have other questions related to that if you don't mind. For me, avoiding having to manually place 300 LEDs is worth some effort. Thanks again.

**Reply by jerseyp2** [reply]
Running your code example gives "AttributeError: 'NoneType' object has no attribute 'Placement'". That must be a problem with the way I created my PolarArray.

**Reply by DeepSOIC** [reply]
Seems like getObject(...) returned None, which is what happens if the name is not correct. This bit of code will absolutely need some adjustments to get the names of led objects correctly. Note that what you see in tree view are labels, not names. Names of objects are displayed in the status bar when you highlight them. You can also make the code search objects by labels using the `.getObjectsByLabel(...)[0]` instead of `.getObject(...)`

**Reply by jerseyp2** [reply]
Thanks! To provide a bit more detail of my setup, in FreeCad, I created a sketch with a single horizontal line in it, and attached a placement. Then I created a PolarArray with 300 placements on it from that placement, and it worked well. After importing the Kicad board, I created a compound from the 300 LED models, because I couldn't seem to select all of them and do anything useful in Lattice2 that way. With the Compound, I have two objects, the Compound and the PolarArray, so that's where I've gotten to so far.

**Comment by jerseyp2** [comment]
Hello

I think it's possible. KicadStepUp allows you to move/position the imported models and push them back to Kicad, so there is no need to overwrite the original positions. They can be translated/moved.

I made a test Kicad schematic/layout with 25 LEDs, and imported it into FreeCad. Then I created a Sketch with 4 dots (to keep it simple) on a circle, and assigned kites to them using Lattice Placement PlacementOfShape: CenterOfMass (Direct Children Only). I then selected a KicadStepUp LED STEP model, and used the Part Workbench Attachment to select a kite's origin vertex, and when I select Translate Origin in the menu, the model snaps to the kite. I did that four times to assign four models to four kites on the circumference of the circle. Then I used KicadStepUp to push the revised Kicad model positions back to Kicad, and the footprints in the Kicad layout are updated. 

In theory, I should be able to write python code to handle my 300 LEDs this way. Thanks for your code sample. It really helped.

![FreeCadLattice2](https://user-images.githubusercontent.com/108946307/178131322-3463ad3f-201e-4fb8-9101-314b7419d514.png)

![KicadLayout](https://user-images.githubusercontent.com/108946307/178131330-5789d603-649f-4611-a20e-a64ad552cf57.png)
